### PR TITLE
BDOG-2530 shared teams fragment

### PR DESCRIPTION
**Pull request by fionahiklas**
Implementing fixes for improvement 3 from https://jira.tools.tax.service.gov.uk/browse/BDOG-2530

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,6 +11,20 @@
 
 See [catalogue-acceptance-tests]("https://github.com/hmrc/catalogue-acceptance-tests") for help setting up dependent services.
 
+Run docker with the following command
```
Comment (colin-lamed) [inline README.md:14]: This is already documented in the above `catalogue-acceptance-tests`, maybe could just reword "for help setting up dependent services and mongo"

File: app/uk/gov/hmrc/cataloguefrontend/connector/TeamsAndRepositoriesConnector.scala [hunk 2]
```diff
@@ -173,9 +173,6 @@ case class GitRepository(
 ) {
   def isShared: Boolean =
```
Comment (colin-lamed) [inline app/uk/gov/hmrc/cataloguefrontend/connector/TeamsAndRepositoriesConnector.scala:174]: `isShared` is only used by the partial - it could be moved into the partial, along with the `Constant.sharedRepoTeamsCutOff`.

File: app/views/partials/GitRepositoryTeamNames.scala.html [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ * limitations under the License.
+ *@
+
+@import uk.gov.hmrc.cataloguefrontend.connector.GitRepository
+@import uk.gov.hmrc.cataloguefrontend.connector.model.TeamName
+
+@this()
+
+@(
+        repository: GitRepository
```
Comment (colin-lamed) [inline app/views/partials/GitRepositoryTeamNames.scala.html:23]: We have two partials that do the same, to support different input types.
Can we change this to take `Seq[Team]` rather than `GitRepository` or `ServiceWithDependency`, then we only need one.
The `isShared` function can be inlined here. Inlining the `Constant.sharedRepoTeamsCutOff` will make it clear that the cutoff is `8`, but the `teamsNames.take` is `15` - which don't align. (`splitAt(8)` and checking if the second list is empty might avoid defining the value more than once)

File: app/views/partials/GitRepositoryTeamNames.scala.html [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+)
+
+@if(!repository.isShared) {
+    @repository.teamNames.map {tn => <div  class="repo-team" ><a href="@uk.gov.hmrc.cataloguefrontend.teams.routes.TeamsController.team(TeamName(tn))">@tn</a></div> }
+} else {
+    <div title="@repository.teamNames.take(15).mkString("\n")" class="repo-team" >
+        <em><a href="@uk.gov.hmrc.cataloguefrontend.routes.CatalogueController.repository(repository.name)#teams">Shared by @repository.teamNames.length teams</a></em>
```
Comment (colin-lamed) [inline app/views/partials/GitRepositoryTeamNames.scala.html:30]: I think we can leave this link out. It might be unexpected to the client to be taken to a different page when clicking on `Shared by ...` - it stands out more when not a link. There's usually a link already available on the repository name in the same table.

File: test/uk/gov/hmrc/cataloguefrontend/JsonData.scala [hunk 5]
```diff
@@ -22,78 +22,21 @@ import uk.gov.hmrc.cataloguefrontend.shuttering.{ShutterStatusValue, ShutterType
 import java.time.Instant
 
 /**
-  * Created by armin.
-  */
+ * Created by armin.
+ *
+ * TODO: Refactor the data here into files under the
```
Comment (colin-lamed) [inline test/uk/gov/hmrc/cataloguefrontend/JsonData.scala:27]: Can these TODO comments be removed now?
You can also remove the "Created by" comment while we're here.

File: app/uk/gov/hmrc/cataloguefrontend/connector/TeamsAndRepositoriesConnector.scala (hunk 2 above)
Comment (colin-lamed) [reply app/uk/gov/hmrc/cataloguefrontend/connector/TeamsAndRepositoriesConnector.scala:174]: Actually, it looks like `prcommenter/model.scala` also duplicates this - can this also use the new partial?

File: README.md (hunk 1 above)
Comment (fionahiklas) [reply README.md:14]: I think this might be backwards: we should have some details about how to run/test the catalogue-frontend locally in this README rather than in the acceptance tests.  Maybe a wider discussion for later about making development easier :)

File: app/views/partials/GitRepositoryTeamNames.scala.html (hunk 4 above)
Comment (fionahiklas) [reply app/views/partials/GitRepositoryTeamNames.scala.html:30]: The link is deliberate to make the table accessible and consistent.  For example, if someone uses a reader that zooms into part of the screen where the teams column is visible it's not ideal that they then have to scroll sideways to get to the link.

File: README.md (hunk 1 above)
Comment (fionahiklas) [reply README.md]: Removed duplicate docs

File: app/uk/gov/hmrc/cataloguefrontend/connector/TeamsAndRepositoriesConnector.scala (hunk 2 above)
Comment (fionahiklas) [reply app/uk/gov/hmrc/cataloguefrontend/connector/TeamsAndRepositoriesConnector.scala:174]: Updated `prcommenter/model.scala` and added tests for the PrCommenter page

File: app/views/partials/GitRepositoryTeamNames.scala.html (hunk 3 above)
Comment (fionahiklas) [reply app/views/partials/GitRepositoryTeamNames.scala.html]: The `GitRepository` object still needs it's `isShared` function and it's used in quite alot of code.  I can raise another ticket to handle this OR we can do it under this ticket?

File: README.md [hunk 6]
```diff
@@ -3,13 +3,19 @@
...
 * [License](#license)
 
 ## Setup
 
-See [catalogue-acceptance-tests]("https://github.com/hmrc/catalogue-acceptance-tests") for help setting up dependent services.
+For help setting up dependent services see [catalogue-acceptance-tests]("https://github.com/hmrc/catalogue-acceptance-tests")
+
+## Development
+
+In order to run unit tests against any changes to views or partials you need to execute
```
Comment (colin-lamed) [inline README.md:17]: `sbt test` already invokes `compile` - did you mean `sbt clean`?

File: app/uk/gov/hmrc/cataloguefrontend/connector/TeamsAndRepositoriesConnector.scala (hunk 2 above)
Comment (colin-lamed) [reply app/uk/gov/hmrc/cataloguefrontend/connector/TeamsAndRepositoriesConnector.scala:174]: Can `isShared` be removed from here now?

File: app/views/partials/TeamNames.scala.html [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+@this()
+
+@(
+        thingLinkFunction: (String) => Call,
+        thingName: String,
+        teamNames: Seq[Any]
+)
+
+@if(teamNames.length < Constant.sharedRepoTeamsCutOff ) {
```
Comment (colin-lamed) [inline app/views/partials/TeamNames.scala.html:28]: If the constant is only required in this file, then it could be inlined here.

File: app/views/partials/TeamNames.scala.html [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+ *@
+
+@import uk.gov.hmrc.cataloguefrontend.connector.model.TeamName
+@import uk.gov.hmrc.cataloguefrontend.config.Constant
+
+@this()
+
+@(
+        thingLinkFunction: (String) => Call,
+        thingName: String,
```
Comment (colin-lamed) [inline app/views/partials/TeamNames.scala.html:24]: A better name would be `repositoryName`

File: app/views/partials/TeamNames.scala.html [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+ * limitations under the License.
+ *@
+
+@import uk.gov.hmrc.cataloguefrontend.connector.model.TeamName
+@import uk.gov.hmrc.cataloguefrontend.config.Constant
+
+@this()
+
+@(
+        thingLinkFunction: (String) => Call,
```
Comment (colin-lamed) [inline app/views/partials/TeamNames.scala.html:23]: It looks like this param exists to support both `repository` and `service` urls. I think using the `repository` endpoint for all cases should be fine, so the reverse route will not need to be passed as a param. (service endpoint only exists for the edge case where the service name doesn't match the repo name, but this is rare, so don't think it's worth over engineering here)

File: test/uk/gov/hmrc/cataloguefrontend/JsonData.scala (hunk 5 above)
Comment (colin-lamed) [reply test/uk/gov/hmrc/cataloguefrontend/JsonData.scala:27]: Or just keep the TODOs locally

File: app/views/partials/GitRepositoryTeamNames.scala.html (hunk 3 above)
Comment (colin-lamed) [reply app/views/partials/GitRepositoryTeamNames.scala.html]: OK - we can leave `isShared`

File: README.md (hunk 6 above)
Comment (fionahiklas) [reply README.md:17]: Sorry, I should have clarified that comment - "run unit tests in the IDE" ... maybe there is an IDE setting but OOTB running a unit test (right click on it in the IDE) should run the required steps for converting partials to scala and then compiling that.  I couldn't see a task that did just the work of partial -> scala so the only solution seemed to be compile.  That seemed to work and appeared to be incremental anyway so not too time consuming.  Maybe there is a way to get the IDE to do this?  Or maybe the advice should be in the README so we don't have to get into IDE config or how people develop code

File: test/uk/gov/hmrc/cataloguefrontend/JsonData.scala (hunk 5 above)
Comment (fionahiklas) [reply test/uk/gov/hmrc/cataloguefrontend/JsonData.scala:27]: The point for the TODOs was because they need to be TOdone :D And it might not always be me TODO'ing them so a local copy wouldn't really help much.  I've added https://jira.tools.tax.service.gov.uk/browse/BDOG-2593 so that the work to tidy up isn't lost

File: app/views/partials/TeamNames.scala.html (hunk 9 above)
Comment (fionahiklas) [reply app/views/partials/TeamNames.scala.html:23]: I've left it parameterized because the table in each page has columns that go to different pages (repository or service), from a usability perspective we should probably stay consistent rather than have the link go to the repository in the case where other links go to service.

I guess if we refactor the code in the future so that the service page/link doesn't exist anymore then we can probably drop the distinction and the parameter.

File: app/uk/gov/hmrc/cataloguefrontend/connector/model/model.scala [hunk 10]
```diff
@@ -23,7 +23,9 @@ import play.api.mvc.{PathBindable, QueryStringBindable}
 
 case class Username(value: String) extends AnyVal
 
-case class TeamName(asString: String) extends AnyVal
+case class TeamName(asString: String) extends AnyVal {
+  override def toString: String = asString
```
Comment (colin-lamed) [inline app/uk/gov/hmrc/cataloguefrontend/connector/model/model.scala:27]: Can we keep `toString` as the default. For logging, debugging, reading test failures etc. it's useful to see the type. This also aligns with all our other Stringable value classes.

File: app/views/partials/TeamNames.scala.html [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+ * limitations under the License.
+ *@
+
+@import uk.gov.hmrc.cataloguefrontend.connector.model.TeamName
+@import uk.gov.hmrc.cataloguefrontend.config.Constant
+
+@this()
+
+@(
+        controllerLinkString: String,
```
Comment (colin-lamed) [inline app/views/partials/TeamNames.scala.html:23]: The parameter name doesn't need to include the type (that info is already available). The name doesn't make it clear what it's used for - I'd suggest `sharedTeamsLink`

File: app/views/partials/TeamNames.scala.html [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+        teamNames: Seq[Any]
```
Comment (colin-lamed) [inline app/views/partials/TeamNames.scala.html:24]: Can the parameter order be swapped? `teamNames` is the primary param, the rest being auxiliary (`sharedTeamsLink` could even have been optional with a default value of `None`)

File: app/views/partials/TeamNames.scala.html [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+)
+
+@if(teamNames.length < Constant.sharedRepoTeamsCutOff ) {
+    @teamNames.map {tn => <div  class="repo-team" ><a href="@uk.gov.hmrc.cataloguefrontend.teams.routes.TeamsController.team(TeamName(tn.toString()))">@tn.toString()</a></div> }
+} else {
+    <div title="@teamNames.take(Constant.sharedRepoTeamsCutOff).mkString("\n")" class="repo-team" >
+        <em><a href="@controllerLinkString">Shared by @teamNames.length teams</a></em>
+    </div>
+}
```
Comment (colin-lamed) [inline app/views/partials/TeamNames.scala.html:33]: IDE can be configured to always add `\n` at EOF

File: test/uk/gov/hmrc/cataloguefrontend/JsonData.scala (hunk 5 above)
Comment (colin-lamed) [reply test/uk/gov/hmrc/cataloguefrontend/JsonData.scala:27]: Can they be removed now we have a ticket?

**Review by colin-lamed — APPROVED** [review]

File: app/views/partials/TeamNames.scala.html (hunk 12 above)
Comment (fionahiklas) [reply app/views/partials/TeamNames.scala.html]: Fixed

File: app/views/partials/TeamNames.scala.html (hunk 11 above)
Comment (fionahiklas) [reply app/views/partials/TeamNames.scala.html]: Fixed

File: test/uk/gov/hmrc/cataloguefrontend/JsonData.scala (hunk 5 above)
Comment (fionahiklas) [reply test/uk/gov/hmrc/cataloguefrontend/JsonData.scala:27]: I guess the point they can be removed is when they are TOdone :)
